### ISSUE TITLE
Possible new restrictions added? 

### ISSUE THREAD

**Issue by amhashicorp** [aidansteele/rdsconn#5]
Hi, 

I have been using this code to successfully connect to an RDS instance in a private subnet. I was working on adding some extra functionality but then started to hit some errors that did not correlate with the changes I had made. 

I decided to go back to trying to use the code in this repo which worked originally and found I could have the connection waiting on a port but then when trying to connect I received the following error:
```
Error: : dialing websocket: websocket: bad handshake
```

I decided to go even more basic and ran this command  - 
```
aws ec2-instance-connect open-tunnel \
  --private-ip-address <IP-ADDRESS-WAS-HERE> \
  --instance-connect-endpoint-id "<ENDPOINT-ID-WAS-HERE>" \
  --remote-port 5432 \
  --local-port 5432 \
  --region us-west-2
```

I found I got this output error:
```
Listening for connections on port 5432.
[1] Accepted new tcp connection, opening websocket tunnel.

awscli.customizations.ec2instanceconnect.websocket - ERROR - {"ErrorCode":"InvalidParameter","Message":"The specified RemotePort is not valid. Specify either 22 or 3389 as the RemotePort and retry your request."}

AWS_ERROR_HTTP_WEBSOCKET_UPGRADE_FAILURE: Failed to upgrade HTTP connection to Websocket.
```

This error seems to suggest that the RemotePort options has now been restricted to 3389 (RDP) and 22 (SSH)? 

Has anyone else experienced this?

**Comment by goosefraba** [comment]
Since some hours, I have the same issue. Everything worked before and when I select a database now after using rdsconn proxy command and try to connect to localhost:5432 I get the same error message.

**Comment by aidansteele** [comment]
Yep, I'm seeing this too. It seems that the service is now restricted to standard SSH and RDP ports. I _assume_ that RDS would still work if configured to listen on port 3389 (instead of 5432) but I don't think that's a reasonable thing for me to suggest people do. I'll keep this issue open if people want to subscribe to hear about updates in the future (e.g. if AWS removes the restriction)

**Comment by avoidik** [comment]
have you added any security group(s) to EICE when you've been creating it?

**Comment by goosefraba** [comment]
Yes and I also tried to allow all traffic but still not working.

**Comment by avoidik** [comment]
what a shame, this is an AWS API limitation now, not a network

**Comment by tomkins** [comment]
Amusingly - this project was included in [AWS Week in Review](https://aws.amazon.com/blogs/aws/aws-week-in-review-generative-ai-with-llm-hands-on-course-amazon-sagemaker-data-wrangler-updates-and-more-july-3-2023/), so at least some people at AWS like the idea of it - even if others have locked it down.

**Comment by yevman** [comment]
Instead of changing the DB sport is there any simple and cheap AWS service that can be used to switch port traffic before hitting the DB from the EIC?

**Comment by aidansteele** [comment]
@yevman I'm not sure it meets the "cheap" criteria, but you could use a Network Load Balancer in TCP mode. It could listen on port 3389 and forward to a target listening on a different port (e.g. your RDS instance listening on port 5432)

**Comment by hector-able** [comment]
I faced the same issue with a different port and used two workarounds, keeping `ec2-instance-connect`. Both work depending on what is your setup:

1. `aws ec2-instance-connect ssh --instance-id XXXX --local-forwarding 5432:localhost:5432` . The `local-forwarding` option works for other ports besides 22 and 3389. 

2. Create a double tunnel first with `ec2-instance-connect` and then with `ssh` like this:
```
# You can create a new key or use an existing one
ssh-keygen -t rsa -N "" -f temporary_key
# This will create a tunnel to port 22
aws ec2-instance-connect open-tunnel --local-port 2222 --remote-port 22 --instance-id XXXXX
# This will send our key to the EC2 instance
aws ec2-instance-connect send-ssh-public-key --instance-id XXXX file://temporary_key.pub --instance-os-user YYYYY
# This will create the tunnel with the port you need 
ssh -o StrictHostKeyChecking=no -i vault -p 2222 -N -L 127.0.0.1:5432:localhost:5432 YYYYY@localhost
```

I'm using both options since AWS disabled the option to create the tunnel with other ports and didn't have any issues. The only downside is that `aws ec2-instance-connect ssh` and `ssh` will open a shell, so moving them to the background could be necessary.

**Comment by MStaniaszek1998** [comment]
> @yevman I'm not sure it meets the "cheap" criteria, but you could use a Network Load Balancer in TCP mode. It could listen on port 3389 and forward to a target listening on a different port (e.g. your RDS instance listening on port 5432)

Is it possible to achieve? 
I have set up the Network Load Balancer to route traffic to RDS instance, but I get the following error when I try to connect to Instance Connect Endpoint: 

**Encountered error with websocket: (9, 'Bad file descriptor')**

**Comment by MStaniaszek** [comment]
> @yevman I'm not sure it meets the "cheap" criteria, but you could use a Network Load Balancer in TCP mode. It could listen on port 3389 and forward to a target listening on a different port (e.g. your RDS instance listening on port 5432)

Is it possible to achieve? I tried this approach but it gives me an error 
**Encountered error with websocket: (9, 'Bad file descriptor')**

**Comment by mathewmoon** [comment]
@MStaniaszek When I got that error it was due to a SG rule. The problem with a NLB is that the target IP addresses for the RDS endpoints are subject to change. I've created a solution with a Lambda that runs every minute to check that the target group IP matches the latest result of `dig` but that's a pretty hacky solution and if you're using IaC you have to either let the Lambda handle creating the target group or, if using Terraform, add a `lifecycle.ignore_changes` to the resource so your next run doesn't try to "fix" it.

I was however able to just create a cluster running on port 3389. I wouldn't do that in production though. If they updated the ports without warning who's to say they won't do something to limit the protocol that is allowed to pass through it. It peeves me off that such a breaking change, in however short a time from initial release, wouldn't be predicated on a proper deprecation notice.

In the end I would spend the extra 5 minutes to create a nano bastion host for SSM before trying to hack around the breaking change. It's a pity though. I was excited about the service.

**Comment by nitrocode** [comment]
I guess it's back to a bastion host.

Maybe [BohdanPetryshyn/basti](https://github.com/BohdanPetryshyn/basti) (temp bastion host + ssm) can be used while we wait for aws to open up port 5432 again via ec2 instance connect?
